pkg/lifecycle/startup/autodetection/autodetection_methods.go

```diff
@@ -0,0 +1,233 @@
+// detection method.
+func AutoDetectCIDR(method string, version int, k8sNode *v1.Node, defInterfacesToExclude []string, getInterfaces func([]string, []string, int) ([]Interface, error)) *cnet.IPNet {
+	if method == "" || method == AUTODETECTION_METHOD_FIRST {
+		// Autodetect the IP by enumerating all interfaces (excluding
+		// known internal interfaces).
+		return autoDetectCIDRFirstFound(version, defInterfacesToExclude)
+	} else if strings.HasPrefix(method, AUTODETECTION_METHOD_INTERFACE) {
+		// Autodetect the IP from the specified interface.
+		ifStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_INTERFACE)
+		// Regexes are passed in as a string separated by ","
+		ifRegexes := regexp.MustCompile(`\s*,\s*`).Split(ifStr, -1)
+		return autoDetectCIDRByInterface(ifRegexes, version)
+	} else if strings.HasPrefix(method, AUTODETECTION_METHOD_CIDR) {
+		// Autodetect the IP by filtering interface by its address.
+		cidrStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_CIDR)
+		// CIDRs are passed in as a string separated by ","
+		matches := []cnet.IPNet{}
+		for _, r := range regexp.MustCompile(`\s*,\s*`).Split(cidrStr, -1) {
+			_, cidr, err := cnet.ParseCIDR(r)
+			if err != nil {
+				log.Errorf("Invalid CIDR %q for IP autodetection method: %s", r, method)
+				return nil
+			}
+			matches = append(matches, *cidr)
+		}
+		return autoDetectCIDRByCIDR(matches, version)
+	} else if strings.HasPrefix(method, AUTODETECTION_METHOD_CAN_REACH) {
+		// Autodetect the IP by connecting a UDP socket to a supplied address.
+		destStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_CAN_REACH)
+		return autoDetectCIDRByReach(destStr, version)
+	} else if strings.HasPrefix(method, AUTODETECTION_METHOD_SKIP_INTERFACE) {
+		// Autodetect the Ip by enumerating all interfaces (excluding
+		// known internal interfaces and any interfaces whose name
+		// matches the given regexes).
+		ifStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_SKIP_INTERFACE)
+		// Regexes are passed in as a string separated by ","
+		ifRegexes := regexp.MustCompile(`\s*,\s*`).Split(ifStr, -1)
+		return autoDetectCIDRBySkipInterface(ifRegexes, version, defInterfacesToExclude)
+	} else if strings.HasPrefix(method, K8S_INTERNAL_IP) {
+		// K8s InternalIP configured for node is used
+		if k8sNode == nil {
+			log.Error("Cannot use method 'kubernetes-internal-ip' when not running on a Kubernetes cluster")
+			return nil
+		}
+		return autoDetectUsingK8sInternalIP(version, k8sNode, getInterfaces)
+	}
+
+	// The autodetection method is not recognised and is required.  Exit.
+	log.Errorf("Invalid IP autodetection method: %s", method)
+	utils.Terminate()
+	return nil
+}
+
+// autoDetectCIDRFirstFound auto-detects the first valid Network it finds across
+// all interfaces (excluding common known internal interface names).
+func autoDetectCIDRFirstFound(version int, defInterfacesToExclude []string) *cnet.IPNet {
```

Comment:
This function takes a new argument now - is this really needed for this PR?

Reply:
yes removed it.